runtime/core/exec_aten/util/tensor_util.cpp

```diff
@@ -0,0 +1,44 @@
+#include "tensor_util.h"
+
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <executorch/runtime/core/exec_aten/util/tensor_util.h>
+
+#include <executorch/runtime/platform/assert.h>
+
+namespace executorch::runtime {
+/**
+ * Shared implementation for tensor_util.h, may only contain code that
+ * works whether or not ATen mode is active.
+ */
+std::array<char, kTensorShapeStringSizeLimit> tensor_shape_to_c_string(
+    executorch::aten::ArrayRef<executorch::aten::SizesType> shape) {
+  std::array<char, kTensorShapeStringSizeLimit> out;
+  char* p = out.data();
+  *p++ = '(';
+  for (const auto elem : shape) {
+    if (elem < 0 || elem > kMaximumPrintableTensorShapeElement) {
+      strcpy(p, "ERR, ");
+      p += strlen("ERR, ");
```

Comment:
Is there a guarantee that p has at least 6 bytes available? `strncpy` would be safer

Reply:
yes, there is a guarantee. we have enough space for a number, so we have enough space for 6 bytes. can add a static_assert about the number of digits in kMaximumPrintableTensorShapeElement

Reply:
Nothing in this function seems to check that shape.size() < kTensorDimensionLimit.

...though looking around the codebase I see that there's a ton of code that assumes that limit without checking, which is pretty terrifying. That constant was originally a little helper value for logging, and afaik the original users of it checked that the dim was smaller before using it. Looks like the kernels have inherited an assumption that the constant is some kind of guarantee, which it is not: the runtime can legally deserialize tensors with a larger number of dims than that value.